entrypoint.sh

```diff
@@ -1,42 +1,48 @@
 #!/bin/bash
 
-dataset_f=$1
-participant_f=$2
-out=${3-.}
-jobs=${4-6}
+machine_f=$1
+out=${2-.}
+jobs=${3-6}
 
 getit() {
-    datalad get --dataset $(cut  -d " " -f1 <<< $1) $(cut -d " " -f2 <<< $1)
+    datalad get --dataset $(cut -d " " -f1 <<< $1) $(cut -d " " -f2 <<< $1)
 }
 export -f getit
 
 [ ! -e ${out} ] && mkdir -p ${out}
 
+installed_datasets=()
+
 {
-    # We need to read the first line and discard it to skip the header row
+    # Skip the header row of the machine_f file
     read
-    # Ensure last line is read even if file does not end with newline
-    # See: https://stackoverflow.com/a/12916758
-    while read -r dataset || [ -n "$dataset" ];
-    do
-
-        ds_id=$(cut -f1 <<< $dataset)
-        ds_full_name=$(cut -f2 <<< $dataset)
-        ds_url=$(cut -f3 <<< $dataset)
-
-        echo Will now install "'"${ds_full_name}"'" from ${ds_url}.
-
-        (
-            cd ${out}
-            datalad install ${ds_url}
-        )
-
-        for ses in $(grep $ds_id ${participant_f} | cut -f9);
-        do
-            ds_name=$( echo $ses | cut -d "/" -f2)
-            ses_path=$( echo $ses | cut -d "/" -f3-)
-            echo ${out}/${ds_name} ${out}/${ds_name}/${ses_path}
-        done | parallel -j${jobs} --jl ${out}/parallel.log getit ::: 2>&1 | tee ${out}/parallel.outs
+    while read -r dataset || [ -n "$dataset" ]; do
+
+        ds_full_name=$(cut -f1 <<< "$dataset")
+        ds_url=$(cut -f2 <<< "$dataset")
+        ses=$(cut -f5 <<< "$dataset")
+
+        # Check if dataset is already installed
+        if [[ ! " ${installed_datasets[@]} " =~ " ${ds_url} " ]]; then
+            echo "Will now install '${ds_full_name}' from ${ds_url}."
+
+            (
+                cd "${out}"
+                datalad install "${ds_url}"
+            )
+
+            # Mark this dataset as installed
+            installed_datasets+=("${ds_url}")
+        else
+            echo "Dataset '${ds_full_name}' from ${ds_url} is already installed. Skipping."
+        fi
+
+        # Process session paths
+        for ses_path in $ses; do
+            ds_name=$(echo "$ses_path" | cut -d "/" -f2)
+            ses_subpath=$(echo "$ses_path" | cut -d "/" -f3-)
+            echo "${out}/${ds_name} ${out}/${ds_name}/${ses_subpath}"
+        done | parallel -j"${jobs}" --jl "${out}/parallel.log" getit ::: 2>&1 | tee -a "${out}/parallel.outs"
```

Comment:
Since we no longer need this inner for loop, to preserve the ability for parallelization, you may want to move this instead to the outer `while` loop. This should allow us to parallelize over rows of the machine results TSV.

Reply:
Ah, the issue with this would be that your current way of tracking installed datasets using an array `installed_datasets=()` assumes a serial order and may not work if two row-operations are happening in parallel.

Alternatively, I would suggest having two separate loops for installing datasets and then getting sessions, respectively. Your first loop could loop over the unique pairs of `PortalURI` and `DatasetName` columns using `cut` directly, e.g.:
```bash
cut -f1,2 "$dataset_f" | sort | uniq | parallel -j"${jobs}" --jl "${out}/parallel.log" "
  
  <your dataset installation logic here>

" ::: 2>&1 | tee -a "${out}/parallel.outs"
```

to first install all the unique datasets. then, your second one would be the `while` loop you already have, but now is just concerned with getting the sessions

Reply:
to clarify: if you go with the two-loops option, both could be parallelized separately